text/0000-placement-by-return.md

```diff
@@ -0,0 +1,926 @@
+  local_0: &'static str,
+}
+impl __MyFunction__Internal {
+  /* "unsize coroutines" are different from normal generators because their behavior is much more
+    restricted. They yield exactly one Layout, and then they are finished by writing to a pointer. */
+  unsafe fn start(state: *mut __MyFunction__Internal) -> Layout {
+    /* As usual for generators, locals (including anonymous ones) get desugared into member
+      variables of an anonymous struct. In this case, the local variable is just a string literal. */
+    state.local_0 = "hello world";
+    Layout::for_value("hello world")
+  }
+  unsafe fn finish(state: *mut __MyFunction__Internal, slot: *mut u8) -> &mut str {
+    ptr::copy(state.local_0, slot.as_mut_ptr(), mem::size_of_val(state.local_0));
+    /* Notice how I also have to return a properly-built fat pointer?
+       This isn't very important for string slices, but it's mandatory for trait objects, because I need to
+       supply a vtable. */
+    str::from_raw_parts_mut(slot, state.local_0.len())
+  }
+}
+
+// desugar function_that_calls_my_function
+use std::alloc::Layout;
+struct __FunctionThatCallsMyFunction__Internal {
+  delegate: __MyFunction__Internal,
+}
+impl __FunctionThatCallsMyFunction__Internal {
+  unsafe fn start(state: *mut __FunctionThatCallsMyFunction__Internal ) -> Layout {
+    /* first, run everything leading up to the return */
+    println!("Hi, there!");
+    /* then, desugar the return (in this case, by delegating */
+    __MyFunction__Internal::start(&mut state.delegate)
+  }
+  unsafe fn finish(state: *mut __FunctionThatCallsMyFunction__Internal, slot: *mut u8) -> &mut str {
+    __MyFunction__Internal::start(&mut state.delegate, slot)
+  }
+}
+```
+
+This interface ends up putting some pretty harsh limitations on what functions that return unsized types can do. An unsized type-returning function can always return the following kinds of expression:
+
+* Constants and dereferencing constants (as shown above). These are desugared by yielding the layout of the literal and returning the literal by copying it.
+* Directly returning the value of another function that also returns the same unsized type. These are desugared by forwarding, as shown above with `function_that_calls_my_function`.
+* Variable-length array literals, similar to those in [RFC 1909](https://github.com/rust-lang/rfcs/blob/master/text/1909-unsized-rvalues.md). These are desugared by yielding the length variable, then returning the payload through ptr offsets and ptr writes.
+* Unsized coercions. These are desugared by storing the sized type as function state, yielding the layout of the sized type, and returning by copying.
+* Blocks, unsafe blocks, and branches that have acceptable expressions in their tail position.
+
+As is typical for generators, these functions may need to be desugared into simple "state machines" if they return branches or have more than one exit point.
+
+```rust
+fn with_branch() -> Option<Error> {
```

Comment:
```suggestion
fn with_branch() -> Option<dyn Error> {
```

BTW I don't think placing a DST inside an enum is going to be supported soon. Could you choose another type?